Drop PHP 7, add PHP 8.2 Support

rogervila opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Support PHP 8.2

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

rogervila committed:
Update composer.json

Signed-off-by: Roger Vilà <rogervila@me.com>

rogervila commented:
Needs laminas-api-tools/api-tools-api-problem to be upgraded (https://github.com/laminas-api-tools/api-tools-api-problem/pull/24)

Ocramius commented on composer.json:
```diff
@@ -29,7 +29,7 @@
     },
     "require": {
-        "php": "^7.3 || ~8.0.0 || ~8.1.0",
+        "php": "^7.3 || ~8.0.0 || ~8.1.0 || ~8.2.0",
```
Needs `composer.lock` update too - possibly good idea to drop PHP 7 to simplify the matrix

rogervila replied:
Done

rogervila committed:
Drop PHP 7, Support PHP 8.2

Signed-off-by: Roger Vilà <rogervila@me.com>

rogervila commented:
Needs `laminas-api-tools/api-tools-content-negotiation` to support PHP 8.2

rogervila commented:
Depends on https://github.com/laminas-api-tools/api-tools-content-negotiation/pull/27

Ocramius commented on psalm.xml.dist:
```diff
@@ -28,6 +28,33 @@
                 <referencedMethod name="PHPUnit\Framework\MockObject\Builder\InvocationMocker::with"/>
             </errorLevel>
         </InternalMethod>
+
```
Let's please baseline these: they should go away on future patches

weierophinney replied:
I've done this now. Fun fact: neither of those legacy classes are referenced anywhere, so the setting had no effect anyways. :smile:

Ocramius commented on src/DbConnectedResource.php:
```diff
@@ -42,6 +42,7 @@ public function create($data)
     {
         $data = $this->retrieveData($data);
         $this->table->insert($data);
+        /** @var int|string $id */
```
These should be inferred from the dependency: let's leave the baseline larger here instead of providing hints

weierophinney replied:
Done.

Ocramius commented on src/DbConnectedResource.php:
```diff
@@ -97,6 +98,7 @@ public function fetch($id)
         if (0 === $resultSet->count()) {
             throw new DomainException('Item not found', 404);
         }
+        /** @psalm-var array|object */
```
Same here: let's leave the baseline larger here, rather than providing inline hints

weierophinney replied:
Done.

Ocramius commented on src/DbConnectedResource.php:
```diff
@@ -128,6 +130,7 @@ protected function retrieveData($data)
             return $filter->getValues();
         }
 
+        /** @psalm-var array<string, mixed> */
```
Isn't `getValues()` already in this shape? 🤔

weierophinney replied:
It is. I've removed the hint, though this means stuffing something into the baseline. But Psalm _should_ have been able to infer this.

Ocramius commented on test/ApplicationTest.php:
```diff
@@ -22,6 +23,12 @@ class ApplicationTest extends TestCase
 {
     use ProphecyTrait;
 
+    /** @var ContainerInterface|ObjectProphecy */
```
We can use native types in here, if needed, otherwise all good, so this is just a nit

weierophinney replied:
Done.

Ocramius reviewed (commented):
Provided an initial review pass: it would be good to keep a larger baseline at first.

Patch is OK otherwise

weierophinney committed:
qa: bump api-tools-hal and api-tools-content-negotiation

Bumps to versions known to test against 8.2.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: do not have Psalm exclude legacy Apigility classes

Turns out, there's no need anyways.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: do not hint stuff we can infer from a dependency

Per @Ocramius

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: use native types instead of annotations

Per @Ocramius

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Currently blocked by laminas-api-tools/api-tools-content-validation#22; turning to that next.

weierophinney committed:
qa: bump api-tools-content-validation to 1.12+

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Now blocked by api-tools-mvc-auth, turning to that.

weierophinney committed:
qa: bump to mvc-auth and oauth2 versions that support 8.2

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: bump api-tools-rest to 1.8+

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: bump rpc and servicemanager deps

- Bumps rpc to 1.8
- Bumps servicemanager to 3.11+
- Drops laminas-zendframework-bridge

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: bump api-tools-versioning to 1.7

Picks up PHP 8.2 support

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
Ready for final review!

Ocramius reviewed (approved):
Thanks @weierophinney
